llama-index-integrations/postprocessor/llama-index-postprocessor-nvidia-rerank/llama_index/postprocessor/nvidia_rerank/base.py

```diff
@@ -0,0 +1,214 @@
+from typing import Any, List, Optional, Literal, Generator
+
+from urllib.parse import urlparse
+from llama_index.core.bridge.pydantic import Field, PrivateAttr, BaseModel
+from llama_index.core.callbacks import CBEventType, EventPayload
+from llama_index.core.postprocessor.types import BaseNodePostprocessor
+from llama_index.core.schema import NodeWithScore, QueryBundle
+import requests
+
+from llama_index.core.base.llms.generic_utils import get_from_param_or_env
+
+
+DEFAULT_MODEL = "nv-rerank-qa-mistral-4b:1"
+DEFAULT_BASE_URL = "https://ai.api.nvidia.com/v1"
+
+
+class Model(BaseModel):
+    id: str
+
+
+class NVIDIARerank(BaseNodePostprocessor):
+    """NVIDIA's API Catalog Reranker Connector."""
+
+    class Config:
+        validate_assignment = True
+
+    model: Optional[str] = Field(
+        default=DEFAULT_MODEL,
+        description="The NVIDIA API Catalog reranker to use.",
+    )
+    top_n: Optional[int] = Field(
+        default=5,
+        ge=0,
+        description="The number of nodes to return.",
+    )
+    max_batch_size: Optional[int] = Field(
+        default=64,
+        ge=1,
+        description="The maximum batch size supported by the inference server.",
+    )
+    _api_key: str = PrivateAttr("API_KEY_NOT_PROVIDED")  # TODO: should be SecretStr
+    _mode: str = PrivateAttr("nvidia")
+    _base_url: str = PrivateAttr(DEFAULT_BASE_URL)
+
+    def _set_api_key(self, nvidia_api_key: str = None, api_key: str = None) -> None:
+        self._api_key = get_from_param_or_env(
+            "api_key",
+            nvidia_api_key or api_key,
+            "NVIDIA_API_KEY",
+            "API_KEY_NOT_PROVIDED",
+        )
+
+    def __init__(
+        self,
+        nvidia_api_key: Optional[str] = None,
+        api_key: Optional[str] = None,
+        **kwargs: Any,
+    ):
+        super().__init__(**kwargs)
+
+        self._set_api_key(nvidia_api_key, api_key)
+
+    @property
+    def available_models(self) -> List[Model]:
+        """Get available models."""
+        # there is one model on ai.nvidia.com and available as a local NIM
+        ids = [DEFAULT_MODEL]
+        return [Model(id=id) for id in ids]
+
+    def mode(
```

Comment:
yea again, this feels like a function that isn't needed? It could be just part of the init right?

Reply:
i agree, this could be part of init. there's low potential for a user to want to create an instance and then mode switch it more than once. this is the design flow we're using consistently across communities. it's also up for review.